### ISSUE TITLE
make demux partitioning (Rust match)

### ISSUE THREAD

**Issue by jhellerstein** [hydro-project/hydroflow#725]
must partition input
any copying etc can be done with tees/merges

**Comment by MingweiSamuel** [comment]
Idea:
instead of calling `my_port_name.push()` which can allow the item to be cloned, etc, instead have the closure return `my_port_name` which is secretly a usize index (maybe) and use that to determine which output port to push to

**Comment by tiemobang** [comment]
How would hash partitioning look like in that case?
Something like the following may benefit from syntactic sugar:

```
switch(|x, var_args!(fzbz, fizz, buzz)| {
let num_ports = 3;
let h = hash(x) % num_ports;
match h {
 0 => fzbz;
 1 => fizz;
 2 => buzz;
 }
})
```
For partitioning the following may be a good alternative:
```
switch_op = switch(|x, out[3]| {
let h = hash(x) % out.len();
out[h]
});
fzbz = switch_op[0];
fizz = switch_op[1];
buzz = switch_op[2];
```

**Comment by MingweiSamuel** [comment]
- no big closure
- enum-based routing
- Tiemo: arithmetic-based routing (hash partitioning)

**Comment by jhellerstein** [comment]
resolved in #911
